src/goto-programs/validate_goto_model.cpp

```diff
@@ -0,0 +1,251 @@
+
+  if(validation_options.check_called_functions)
+    check_called_functions();
+
+  if(validation_options.check_sourcecode_location)
+    check_sourcecode_location();
+}
+
+void validate_goto_modelt::entry_point_exists()
+{
+  DATA_CHECK(
+    function_map.find(goto_functionst::entry_point()) != function_map.end(),
+    "an entry point must exist");
+}
+
+void validate_goto_modelt::function_pointer_calls_removed()
+{
+  for(const auto &fun : function_map)
+  {
+    for(auto &instr : fun.second.body.instructions)
+    {
+      if(instr.is_function_call())
+      {
+        const code_function_callt &function_call =
+          to_code_function_call(instr.code);
+        DATA_CHECK(
+          function_call.function().id() == ID_symbol,
+          "no calls via function pointer should be present");
+      }
+    }
+  }
+}
+
+void validate_goto_modelt::check_returns_removed()
+{
+  for(const auto &fun : function_map)
+  {
+    const goto_functiont &goto_function = fun.second;
+    DATA_CHECK(
+      goto_function.type.return_type().id() == ID_empty,
+      "functions must have empty return type");
+
+    for(const auto &instr : goto_function.body.instructions)
+    {
+      DATA_CHECK(
+        !instr.is_return(), "no return instructions should be present");
+
+      if(instr.is_function_call())
+      {
+        const auto &function_call = to_code_function_call(instr.code);
+        DATA_CHECK(
+          function_call.lhs().is_nil(), "function call return should be nil");
+
+        const auto &callee = to_code_type(function_call.function().type());
+        DATA_CHECK(callee.return_type().id() == ID_empty, "");
+      }
+    }
+  }
+}
+
+void validate_goto_modelt::check_called_functions()
+{
+  class test_for_function_addresst : public const_expr_visitort
+  {
+  public:
+    std::set<irep_idt> identifiers;
+
+    void operator()(const exprt &expr) override
+    {
+      if(expr.id() == ID_address_of)
+      {
+        const exprt &pointee = to_address_of_expr(expr).object();
+        if(pointee.id() == ID_symbol)
+          identifiers.insert(to_symbol_expr(pointee).get_identifier());
+      }
+    }
+
+    void clear()
+    {
+      identifiers.clear();
+    }
+  } test_for_function_address;
+
+  for(const auto &fun : function_map)
+  {
+    for(auto &instr : fun.second.body.instructions)
+    {
+      if(instr.is_function_call())
+      {
+        const auto &function_call = to_code_function_call(instr.code);
+        const irep_idt &identifier =
+          to_symbol_expr(function_call.function()).get_identifier();
+
+        DATA_CHECK(
+          function_map.find(identifier) != function_map.end(),
+          "every function call callee must be in the function map");
+      }
+
+      const exprt &src{instr.code};
+      src.visit(test_for_function_address);
+      if(!test_for_function_address.identifiers.empty())
+      {
+        for(auto &identifier : test_for_function_address.identifiers)
+          DATA_CHECK(
```

Comment:
Rather than mixing up the control flow, and having to store the  identifiers to check in an intermediate set (with corresponding house keeping logic to clear the set) - would it be better to pass 'function_map' as a constructor parameter, e.g. something like `src.visit(test_for_function_addresst{function_map})` - then move this `DATA_CHECK` inside `test_for_function_addresst operator()`  - that way, you would not need to maintain an additional set, and could drop the `clear()` member function.

Reply:
Done.